pkg/daemon/pinned_image_set_manager.go

```diff
@@ -0,0 +1,1096 @@
+	if !ok {
+		return fmt.Errorf("node %q has no ephemeral storage capacity", p.nodeName)
+	}
+
+	capacity := storageCapacity.Value()
+	if storageCapacity.Cmp(p.minStorageAvailableBytes) < 0 {
+		return fmt.Errorf("%w capacity: %d, required: %d", errInsufficientStorage, capacity, p.minStorageAvailableBytes.Value())
+	}
+
+	if err := p.checkImagePayloadStorage(ctx, imageSet.Spec.PinnedImages, capacity); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *PinnedImageSetManager) checkImagePayloadStorage(ctx context.Context, images []mcfgv1alpha1.PinnedImageRef, capacity int64) error {
+	// calculate total required storage for all images.
+	requiredStorage := int64(0)
+	for _, image := range images {
+		imageName := image.Name
+		exists, err := p.criClient.ImageStatus(ctx, imageName)
+		if err != nil {
+			return err
+		}
+		if exists {
+			// dont account for storage if the image already exists
+			continue
+		}
+
+		size, err := getImageSize(ctx, imageName, p.authFilePath)
+		if err != nil {
+			return err
+		}
+
+		// account for decompression
+		requiredStorage += size * 2
+	}
+
+	if requiredStorage >= capacity-p.minStorageAvailableBytes.Value() {
+		klog.Errorf("%v capacity=%d, required=%d", errInsufficientStorage, capacity, requiredStorage+p.minStorageAvailableBytes.Value())
+		return errInsufficientStorage
+	}
+
+	return nil
+}
+
+func isImageSetInPool(imageSet string, pool *mcfgv1.MachineConfigPool) bool {
+	for _, set := range pool.Spec.PinnedImageSets {
+		if set.Name == imageSet {
+			return true
+		}
+	}
+	return false
+}
+
+// ensureCrioPinnedImagesConfigFile creates the crio config file which populates pinned_images with a unique list of images then reloads crio.
+// If the list is empty, and a config exists on disk the file is removed and crio is reloaded.
+func (p *PinnedImageSetManager) ensureCrioPinnedImagesConfigFile(images []mcfgv1alpha1.PinnedImageRef) error {
+	hasCrioConfigFile, err := p.hasPinnedImagesConfigFile()
+	if err != nil {
+		return fmt.Errorf("failed to check crio config file: %w", err)
+	}
+
+	// if there are no images and the config file exists, remove it
+	if len(images) == 0 {
+		if hasCrioConfigFile {
+			// remove the crio config file and reload crio
+			if err := p.deleteCrioConfigFile(); err != nil {
+				return fmt.Errorf("failed to remove crio config file: %w", err)
+			}
+			return p.crioReload()
+		}
+		return nil
+	}
+
+	// ensure list elements are unique
+	// assume the number of duplicates is small so preallocate makes sense
+	// for larger lists
+	seen := make(map[string]struct{}, len(images))
+	unique := make([]string, 0, len(images))
+	for _, image := range images {
+		if _, ok := seen[image.Name]; !ok {
+			seen[image.Name] = struct{}{}
+			unique = append(unique, image.Name)
+		}
+	}
+
+	// create ignition file for crio drop-in config. this is done to use the existing
+	// atomic writer functionality.
+	ignFile, err := createCrioConfigIgnitionFile(unique)
```

Comment:
Just a future thought, probably should guard this directory so users can't write to it via machineconfigs and stomp over each other